src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -157,15 +256,299 @@ private static TensorFlowTransform Create(IHostEnvironment env, ModelLoadContext
+                    throw _host.ExceptParam(nameof(args.LearningRateOperation), $"'{args.LearningRateOperation}' does not exist in the model");
+            }
+        }
+
+        private void TrainCore(Arguments args, string model, IDataView input)
+        {
+            var inputsForTraining = new string[Inputs.Length + 1];
+            var inputColIndices = new int[inputsForTraining.Length];
+            var isInputVector = new bool[inputsForTraining.Length];
+            var tfInputTypes = new TFDataType[inputsForTraining.Length];
+            var tfInputShapes = new TFShape[inputsForTraining.Length];
+
+            for (int i = 0; i < Inputs.Length; i++)
+            {
+                inputsForTraining[i] = Inputs[i];
+            }
+
+            inputsForTraining[inputsForTraining.Length - 1] = args.LabeLColumn;
+
+            var inputSchema = input.Schema;
+            for (int i = 0; i < inputsForTraining.Length; i++)
+            {
+                if (!inputSchema.TryGetColumnIndex(inputsForTraining[i], out inputColIndices[i]))
+                    throw _host.Except($"Column {inputsForTraining[i]} doesn't exist");
+
+                var type = inputSchema.GetColumnType(inputColIndices[i]);
+                isInputVector[i] = type.IsVector;
+
+                var tfInput = new TFOutput(Graph[inputsForTraining[i]]);
+                tfInputTypes[i] = tfInput.OutputType;
+                tfInputShapes[i] = Graph.GetTensorShape(tfInput);
+                if (tfInputShapes[i].NumDimensions != -1)
+                {
+                    var newShape = new long[tfInputShapes[i].NumDimensions];
+                    newShape[0] = tfInputShapes[i][0] == -1 ? args.BatchSize : tfInputShapes[i][0];
+
+                    for (int j = 1; j < tfInputShapes[i].NumDimensions; j++)
+                        newShape[j] = tfInputShapes[i][j];
+                    tfInputShapes[i] = new TFShape(newShape);
+                }
+
+                var expectedType = TensorFlowUtils.Tf2MlNetType(tfInputTypes[i]);
+                if (type.ItemType != expectedType)
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", inputsForTraining[i], expectedType.ToString(), type.ToString());
+            }
+
+            var fetchList = new List<string>();
+            if (args.LossOperation != null)
+                fetchList.Add(args.LossOperation);
+            if (args.MetricOperation != null)
+                fetchList.Add(args.MetricOperation);
+
+            var hashedIndices = new HashSet<int>(inputColIndices);
+            for (int epoch = 0; epoch < args.Epoch; epoch++)
+            {
+                using (var cursor = input.GetRowCursor(a => hashedIndices.Contains(a)))
+                {
+                    var srcTensorGetters = GetTensorValueGetters(cursor, inputColIndices, isInputVector, tfInputTypes, tfInputShapes);
+
+                    float loss=0;
+                    float metric=0;
```

Comment:
Do you use these anywhere?

Reply:
yes on line 418 and 419

---
In reply to: [222500966](https://github.com/dotnet/machinelearning/pull/1063#discussion_r222500966) [](ancestors = 222500966)

Reply:
It seems that they are just used to accumulate some values, but the accumulated values are not used anywhere.

---
In reply to: [222504724](https://github.com/dotnet/machinelearning/pull/1063#discussion_r222504724) [](ancestors = 222504724,222500966)

Reply:
These are used on line `400` and `402`. I am working with @ivmatan@microsoft.com to use ProgressChannel more effectively. So, these will be used properly once I do that.

---
In reply to: [222734143](https://github.com/dotnet/machinelearning/pull/1063#discussion_r222734143) [](ancestors = 222734143,222504724,222500966)

Reply:
Is this planned for a separate PR?

---
In reply to: [222755420](https://github.com/dotnet/machinelearning/pull/1063#discussion_r222755420) [](ancestors = 222755420,222734143,222504724,222500966)